#!/bin/bash

#usage: script_chris_contig_assembly_and_mito_extraction_and_taxonomy.sh "Reads_1" "Reads_2" "number_of_blast_hits (I set it to 5 for this purpose)"

exec 3>&1 4>&2
trap 'exec 2>&4 1>&3' 0 1 2 3
exec 1>script_chris_contig_assembly_and_mito_extraction_and_taxonomy_log.txt 2>&1

R1=$1
R2=$2
k=$3

start=`date +%s`

echo -e "\nSTART RUNNING SCRIPT AT $(date)\n\n"

# Running TrimGalore
echo -e "~~~~~~~~~~ RUNNING TRIMGALORE ~~~~~~~~~~\n\n"
mkdir trimgalore_output
trim_galore --fastqc --dont_gzip --paired -o trimgalore_output $R1 $R2

# Running SPAdes
echo -e "\n~~~~~~~~~~ RUNNING SPADES ~~~~~~~~~~\n\n"
mkdir spades_output
name1=$(echo ${R1##*/})
name2=$(echo ${R2##*/})
spades.py -1 ./trimgalore_output/${name1::-6}_val_1.fq -2 ./trimgalore_output/${name2::-6}_val_2.fq -o spades_output

# Shorten contigs names
sed -re 's/(>NODE_[0-9]+)[^=]*$/\1/' ./spades_output/contigs.fasta > ./spades_output/contigs_shorter_names_for_prokka.fasta

# Running prokka
echo -e "\n\n~~~~~~~~~~ RUNNING PROKKA ~~~~~~~~~~\n\n"
prokka --notbl2asn --cpus 4 --outdir prokka_output --prefix prokka_$(echo ${name1%_*}) ./spades_output/contigs_shorter_names_for_prokka.fasta

# Blast prokka proteins translated from contigs (generated by spades) with diamond against diamond nr database in blast format plus taxonomic id plus scientific names
echo -e "\n\n~~~~~~~~~~ RUNNING DIAMOND ~~~~~~~~~~\n\n"
diamond blastp -d ~/Programs/Chris/DB/nr.dmnd -q ./prokka_output/prokka_$(echo ${name1%_*}).faa -o diamond_output.txt --salltitles -k $k --outfmt 6  qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle staxids

# Get nodenames and sequence ids from prokka output
cat ./prokka_output/prokka_$(echo ${name1%_*}).gff | grep -v '##' | sed 's/ID=//g' | sed 's/;/\t/1' | cut -f1,9 > NODE_SEQID.txt

# Merge nodenames and sequence ids on blast output
echo -e "\n\n~~~~~~~~~~ MERGING 1 ~~~~~~~~~~\n"
mergeFilesOnColumn.pl NODE_SEQID.txt diamond_output.txt 2 1 | cut -f-1,3- > NODE_SEQID_diamond_output.txt

# Only use mitochondrial hits
grep mitoch* NODE_SEQID_diamond_output.txt > NODE_SEQID_diamond_output_mitochondrial.txt

# Get contig length and coverage 
cat ./spades_output/contigs.fasta | grep '>' | cut -c 2- | sed 's/_/\t/2' | sed 's/_/\t/2' | sed 's/_/\t/2' | sed 's/_/\t/2' | cut -f1,3,5 > contigs_nodenames_length_coverage.txt

# Merge contig length and coverage on blast output
echo -e "\n\n~~~~~~~~~~ MERGING 2 ~~~~~~~~~~\n"
mergeFilesOnColumn.pl contigs_nodenames_length_coverage.txt NODE_SEQID_diamond_output_mitochondrial.txt 1 1 | cut -f-3,5- > NODE_LENGTH_COVERAGE_SEQID_diamond_output_mitochondrial.txt

# Get taxonomic classification of blasthits 
ete3 ncbiquery --info --search $(cut -f 17 NODE_LENGTH_COVERAGE_SEQID_diamond_output_mitochondrial.txt) >matching_lineages.tsv

# Add taxonomic classification to blastoutput
echo -e "\n\n~~~~~~~~~~ ADDING TAXONOMY CLASSIFICATION ~~~~~~~~~~\n\n"
LookupTaxonDetails2.py -b NODE_LENGTH_COVERAGE_SEQID_diamond_output_mitochondrial.txt -l matching_lineages.tsv -o final_output_mitochondrial_contigs_and_taxonomy.txt -t 17


echo -e "\n\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\nSCRIPT RUNTIME: $((($(date +%s)-$start)/3600))h $((($(date +%s)-$start)%60))m"

# Delete warnings from first merging within log file <-- not sure where warnings come from, command however results in wanted output, warnings therefore ignored and deleted due to massive space in log file
sed -i '/^Use of/ d' script_chris_contig_assembly_and_mito_extraction_and_taxonomy_log.txt

# Sort files
mkdir final_output
mv NODE* matching_lineages.tsv contigs_nodenames_length_coverage.txt diamond_output.txt final_output_mitochondrial_contigs_and_taxonomy.txt final_output/
mkdir script_chris_contig_assembly_and_mito_extraction_and_taxonomy_$(echo ${name1%_*})
mv prokka_output spades_output trimgalore_output final_output script_chris_contig_assembly_and_mito_extraction_and_taxonomy_log.txt script_chris_contig_assembly_and_mito_extraction_and_taxonomy_$(echo ${name1%_*})/